src/waveform/renderers/allshader/waveformrendererrgb.cpp

```diff
@@ -137,63 +137,76 @@ void WaveformRendererRGB::paintGL() {
                 const WaveformData& waveformData = data[i];
 
-                const float filteredLow = static_cast<float>(waveformData.filtered.low);
-                const float filteredMid = static_cast<float>(waveformData.filtered.mid);
-                const float filteredHigh = static_cast<float>(waveformData.filtered.high);
-
-                maxLow = math_max(maxLow, filteredLow);
-                maxMid = math_max(maxMid, filteredMid);
-                maxHigh = math_max(maxHigh, filteredHigh);
-
-                const float all = math_pow2(filteredLow) * lowGain +
-                        math_pow2(filteredMid) * midGain +
-                        math_pow2(filteredHigh) * highGain;
-                maxAll[chn] = math_max(maxAll[chn], all);
+                u8maxLow = math_max(u8maxLow, waveformData.filtered.low);
+                u8maxMid = math_max(u8maxMid, waveformData.filtered.mid);
+                u8maxHigh = math_max(u8maxHigh, waveformData.filtered.high);
+                u8maxAllChn[chn] = math_max(u8maxAllChn[chn], waveformData.filtered.all);
             }
         }
 
+        // Cast to float
+        float maxLow = static_cast<float>(u8maxLow);
+        float maxMid = static_cast<float>(u8maxMid);
+        float maxHigh = static_cast<float>(u8maxHigh);
+        float maxAllChn[2]{static_cast<float>(u8maxAllChn[0]), static_cast<float>(u8maxAllChn[1])};
+
+        // Calculate the magnitude of the maxLow, maxMid and maxHigh values
+        const float magnitude = std::sqrt(
+                math_pow2(maxLow) + math_pow2(maxMid) + math_pow2(maxHigh));
+
+        // Apply the gains
         maxLow *= lowGain;
         maxMid *= midGain;
         maxHigh *= highGain;
 
+        // Calculate the magnitude of the gained maxLow, maxMid and maxHigh values
+        const float magnitudeGained = std::sqrt(
+                math_pow2(maxLow) + math_pow2(maxMid) + math_pow2(maxHigh));
+
+        // The maxAll values will be used to draw the amplitude. We scale them according to
+        // magnitude of the gained maxLow, maxMid and maxHigh values
+        if (magnitude != 0.f) {
+            const float factor = magnitudeGained / magnitude;
+            maxAllChn[0] *= factor;
+            maxAllChn[1] *= factor;
+        }
+
+        // Use the gained maxLow, maxMid and maxHigh values to calculate the color components
         float red = maxLow * low_r + maxMid * mid_r + maxHigh * high_r;
         float green = maxLow * low_g + maxMid * mid_g + maxHigh * high_g;
         float blue = maxLow * low_b + maxMid * mid_b + maxHigh * high_b;
 
-        const float max = math_max3(red, green, blue);
-
-        // Normalize red, green, blue, using the maximum of the three
-
-        if (max == 0.f) {
-            // avoid division by 0
+        // Normalize the color components using the maximum of the three
+        const float maxComponent = math_max3(red, green, blue);
+        if (maxComponent == 0.f) {
```

Comment:
do we have to worry about near-zero values causing the normFactor to explode?

Reply:
no, if you look at how red green and blue are calculated, there is no such risk.